src/garage/torch/algos/vpg.py

```diff
@@ -128,92 +134,149 @@ def train_once(self, itr, paths):
+            actions (torch.Tensor): Actions fed to the environment.
+            rewards (torch.Tensor): Acquired rewards.
+            advantages (torch.Tensor): Advantage value at each step.
+
+        Returns:
+            torch.Tensor: Calculated mean scalar value of loss
+
+        """
+        loss = self._compute_loss_with_adv(obs, actions, rewards, advantages)
+
+        self._optimizer.zero_grad()
+        loss.backward()
+
+        self._optimize(obs, actions, rewards, advantages)
 
-    def _compute_loss(self, itr, obs, actions, rewards, valids, baselines):
+        return loss
+
+    def _compute_loss(self, obs, actions, rewards, valids, baselines):
         """Compute mean value of loss.
 
         Args:
-            itr (int): Iteration number.
             obs (torch.Tensor): Observation from the environment.
-            actions (torch.Tensor): Predicted action.
-            rewards (torch.Tensor): Feedback from the environment.
-            valids (list[int]): Array of length of the valid values.
+            actions (torch.Tensor): Actions fed to the environment.
+            rewards (torch.Tensor): Acquired rewards.
+            valids (list[int]): Numbers of valid steps in each paths.
             baselines (torch.Tensor): Value function estimation at each step.
 
         Returns:
-            torch.Tensor: Calculated mean value of loss
+            torch.Tensor: Calculated negative mean scalar value of objective
 
         """
-        del itr
+        obs_flat = torch.cat(filter_valids(obs, valids))
+        actions_flat = torch.cat(filter_valids(actions, valids))
+        rewards_flat = torch.cat(filter_valids(rewards, valids))
+        advantages_flat = self._compute_advantage(rewards, valids, baselines)
 
-        policy_entropies = self._compute_policy_entropy(obs)
+        return self._compute_loss_with_adv(obs_flat, actions_flat,
+                                           rewards_flat, advantages_flat)
 
-        if self._maximum_entropy:
-            rewards += self._policy_ent_coeff * policy_entropies
+    def _compute_loss_with_adv(self, obs, actions, rewards, advantages):
+        """Compute mean value of loss.
 
+        Args:
+            obs (torch.Tensor): Observation from the environment.
+            actions (torch.Tensor): Actions fed to the environment.
+            rewards (torch.Tensor): Acquired rewards.
+            advantages (torch.Tensor): Advantage value at each step.
+
+        Returns:
+            torch.Tensor: Calculated negative mean scalar value of objective
+
+        """
+        objective = self._compute_objective(advantages, obs, actions, rewards)
+        return -objective.mean()
+
+    def _compute_advantage(self, rewards, valids, baselines):
+        """Compute mean value of loss.
+
+        Args:
+            rewards (torch.Tensor): Acquired rewards.
+            valids (list[int]): Numbers of valid steps in each paths.
```

Comment:
why is valids a list when everything else is a tensor? this seems incongruous

please add shape documentation here